Feature/webdav locking frontend

DeepDiver1975 opened the pull request:
## Description
Frontend part of #31651 

## Related Issue
- refs #31651

## How Has This Been Tested?
- manual testing
- js unit tests are missing

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

## Checklist:
- [x] Code changes
- [x] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:
- [x] unit tests
- [x] properly update status on unlock
- [x] decide on information to be displayed on the details tab

DeepDiver1975 committed:
Handle persistent locks in files app UI

Handle http 423 on file operations
Show lock status to file list in its own sidebar tab

PVince81 commented:
- [ ] requires https://github.com/owncloud/core/pull/31651

The code looks good, we'll need to add unit tests as well.
Do you want me to take over for those ?

I suggest we wait for the backend PR to be merged, unless you can make this PR detect whether the server supports locking :-D

(also something to think about for Phoenix)

Commit:
Improve locks tab + add JS tests

Improve lock tab view code.
Add tests for lock messages in file list
Added tests for the lock notification messages when performing move,
rename or delete.
Add JS tests for file locking plugin
Fixed XML parsing code to work with PhantomJS and potentially others by
using more standard functions.
Added filter to remove text nodes in case of paddings in the XML.
Add missing translation.
Don't show lock icon if lock array is empty
Don't display sidebar lock tab when no locks set
Properly trigger change event for activeLocks
Adjust locked error message for upload

PVince81 commented on apps/files/js/filelockplugin.js:
```diff
@@ -0,0 +1,129 @@
+							locktype: xmlvalue.getElementsByTagName('d:locktype')[0].children[0].localName,
+							lockroot: xmlvalue.getElementsByTagName('d:lockroot')[0].children[0].innerHTML,
+							depth: parseInt(xmlvalue.getElementsByTagName('d:depth')[0].innerHTML, 10),
+							timeout: xmlvalue.getElementsByTagName('d:timeout')[0].innerHTML,
```
at first I though we should parse it to int or float but let's leave it as string to avoid precision issues in the future

PVince81 replied:
turns out the format is "Seconds-X" so it's a string anyway...

PVince81 commented on apps/files/js/filelockplugin.js:
```diff
@@ -0,0 +1,129 @@
+							lockscope: xmlvalue.getElementsByTagName('d:lockscope')[0].children[0].localName,
+							locktype: xmlvalue.getElementsByTagName('d:locktype')[0].children[0].localName,
+							lockroot: xmlvalue.getElementsByTagName('d:lockroot')[0].children[0].innerHTML,
+							depth: parseInt(xmlvalue.getElementsByTagName('d:depth')[0].innerHTML, 10),
```
I've seen here https://datatracker.ietf.org/doc/rfc4918/?include_text=1 (search for "lockroot") that the depth can also be "infinity" @DeepDiver1975

PVince81 replied:
infinity is definitely possible:
```
curl -u user1:test -X LOCK http://localhost/owncloud/remote.php/webdav/test -d "<?xml version='1.0' encoding='UTF-8'?><d:lockinfo xmlns:d='DAV:'> <d:lockscope><d:shared/></d:lockscope></d:lockinfo>" -H "Timeout: Second-120" -H "Depth: infinity"
```

=> `   <d:depth>infinity</d:depth>`

I'll adjust the frontend

PVince81 replied:
ok seems I already adjusted this, marking as resolved

PVince81 commented:
The XML part fails on PhantomJS but not Chrome or Firefox.

I've decided to take this opportunity to switch to ChromeHeadless, needs review: https://github.com/owncloud/core/pull/33296

- [ ] REQUIRES: https://github.com/owncloud/core/pull/33296

PVince81 commented on apps/files/js/locktabview.js:
```diff
@@ -0,0 +1,92 @@
+		'<div class="clear-float mainFileInfoView"></div>' +
+		'{{#each locks}}' +
+		'<div>' +
+		'<div style="display: inline;">{{owner}} has locked this resource via {{lockroot}}</div>' +
```
@DeepDiver1975 is lockroot the full Webdav path like /owncloud/remote.php/dav/... or shortened ?

PVince81 replied:
solved: we are shortening on display

PVince81 commented:
@DeepDiver1975 I'm done with the JS tests.

Had to do some small fixes in the original code, like:

Mind retesting with real data ?
Also please answer my comments above.
Also review my modifications.

Thanks!

PVince81 commented:
@ownclouders rebase

ownclouders commented:
Hey! I'm [GitMate.io](https://gitmate.io)! This pull request is being rebased automatically. Please **DO NOT** push while rebase is in progress or your changes would be lost permanently :warning:

ownclouders commented:
Automated rebase with [GitMate.io](https://gitmate.io) was successful! :tada:

DeepDiver1975 commented:
@PVince81 one thing - which I failed to implement myself - if the user removes the last lock the icon in the file list is not removed. No idea how to do this ....

PVince81 commented:
@DeepDiver1975 I thought this would already work: whenever you change the FileInfoModel by setting attributes, there's a listener in FileList that would rerender the current row and would remove the lock. Maybe it needs more tweaking.

individual-it commented:
- [ ]  bug: Sharing tab enables the lock symbol
![sharingtabbug](https://user-images.githubusercontent.com/2425577/47778155-520ad880-dd1e-11e8-9a8e-a3db4eb832a7.gif)

individual-it commented:
@ownclouders rebase

ownclouders commented:
Hey! I'm [GitMate.io](https://gitmate.io)! This pull request is being rebased automatically. Please **DO NOT** push while rebase is in progress or your changes would be lost permanently :warning:

ownclouders commented:
Automated rebase with [GitMate.io](https://gitmate.io) was successful! :tada:

individual-it committed:
Acceptance tests for webdav locking frontend

individual-it commented:
@phil-davis please review the acceptance tests

PVince81 commented:
@phil-davis ping for review of tests

PVince81 commented:
@ownclouders rebase

ownclouders commented:
Hey! I'm [GitMate.io](https://gitmate.io)! This pull request is being rebased automatically. Please **DO NOT** push while rebase is in progress or your changes would be lost permanently :warning:

ownclouders commented:
Automated rebase with [GitMate.io](https://gitmate.io) was successful! :tada:

PVince81 commented:
@individual-it seems something is broken with the tests setup. It was like this before and after the rebase:
```
  Scenario: setting a lock shows the display name of a user in the locking details                                                       # /drone/src/tests/acceptance/features/webUIWebdavLocks/locks.feature:27
    Given these users have been created:                                                                                                 # FeatureContext::theseUsersHaveBeenCreated()
      | username       |
      | brand-new-user |
      Exception: could not create user. 400 <?xml version="1.0"?>
      <ocs>
       <meta>
        <status>failure</status>
        <statuscode>400</statuscode>
        <message>User already exists</message>
       </meta>
       <data/>
      </ocs>
       in /drone/src/tests/acceptance/features/bootstrap/Provisioning.php:1235
      Stack trace:
      #0 /drone/src/tests/acceptance/features/bootstrap/Provisioning.php(284): FeatureContext->createUser('brand-new-user', '123456', NULL, NULL, true, 'api', false)
      #1 [internal function]: FeatureContext->theseUsersHaveBeenCreated('', '', Object(Behat\Gherkin\Node\TableNode))
      #2 /drone/src/lib/composer/behat/behat/src/Behat/Testwork/Call/Handler/RuntimeCallHandler.php(109): call_user_func_array(Array, Array)
      #3 /drone/src/lib/composer/behat/behat/src/Behat/Testwork/Call/Handler/RuntimeCallHandler.php(64): Behat\Testwork\Call\Handler\RuntimeCallHandler->executeCall(Object(Behat\Behat\Definition\Call\DefinitionCall))
      #4 /drone/src/lib/composer/behat/behat/src/Behat/Testwork/Call/CallCenter.php(140): Behat\Testwork\Call\Handler\RuntimeCallHandler->handleCall(Object(Behat\Behat\Definition\Call\DefinitionCall))
      #5 /drone/src/lib/composer/behat/behat/src/Behat/Testwork/Call/CallCenter.php(96): Behat\Testwork\Call\CallCenter->handleCall(Object(Behat\Behat\Definition\Call\DefinitionCall))
      #6 /drone/src/lib/composer/behat/behat/src/Behat/Behat/Tester/Runtime/RuntimeStepTester.php(125): Behat\Testwork\Call\CallCenter->makeCall(Object(Behat\Behat\Definition\Call\DefinitionCall))
      #7 /drone/src/lib/composer/behat/behat/src/Behat/Behat/Tester/Runtime/RuntimeStepTester.php(73): Behat\Behat\Tester\Runtime\RuntimeStepTester->testDefinition(Object(Behat\Behat\Context\Environment\InitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), Object(Behat\Gherkin\Node\StepNode), Object(Behat\Behat\Definition\SearchResult), false)
      #8 /drone/src/lib/composer/behat/behat/src/Behat/Behat/Hook/Tester/HookableStepTester.php(74): Behat\Behat\Tester\Runtime\RuntimeStepTester->test(Object(Behat\Behat\Context\Environment\InitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), Object(Behat\Gherkin\Node\StepNode), false)
      #9 /drone/src/lib/composer/behat/behat/src/Behat/Behat/EventDispatcher/Tester/TickingStepTester.php(59): Behat\Behat\Hook\Tester\HookableStepTester->test(Object(Behat\Behat\Context\Environment\InitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), Object(Behat\Gherkin\Node\StepNode), false)
      #10 /drone/src/lib/composer/behat/behat/src/Behat/Behat/EventDispatcher/Tester/EventDispatchingStepTester.php(73): Behat\Behat\EventDispatcher\Tester\TickingStepTester->test(Object(Behat\Behat\Context\Environment\InitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), Object(Behat\Gherkin\Node\StepNode), false)
      #11 /drone/src/lib/composer/behat/behat/src/Behat/Behat/Tester/StepContainerTester.php(59): Behat\Behat\EventDispatcher\Tester\EventDispatchingStepTester->test(Object(Behat\Behat\Context\Environment\InitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), Object(Behat\Gherkin\Node\StepNode), false)
      #12 /drone/src/lib/composer/behat/behat/src/Behat/Behat/Tester/Runtime/RuntimeBackgroundTester.php(72): Behat\Behat\Tester\StepContainerTester->test(Object(Behat\Behat\Context\Environment\InitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), Object(Behat\Gherkin\Node\BackgroundNode), false)
      #13 /drone/src/lib/composer/behat/behat/src/Behat/Behat/EventDispatcher/Tester/EventDispatchingBackgroundTester.php(73): Behat\Behat\Tester\Runtime\RuntimeBackgroundTester->test(Object(Behat\Behat\Context\Environment\InitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), false)
      #14 /drone/src/lib/composer/behat/behat/src/Behat/Behat/Tester/Runtime/RuntimeScenarioTester.php(102): Behat\Behat\EventDispatcher\Tester\EventDispatchingBackgroundTester->test(Object(Behat\Behat\Context\Environment\InitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), false)
      #15 /drone/src/lib/composer/behat/behat/src/Behat/Behat/Tester/Runtime/RuntimeScenarioTester.php(70): Behat\Behat\Tester\Runtime\RuntimeScenarioTester->testBackground(Object(Behat\Behat\Context\Environment\InitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), false)
      #16 /drone/src/lib/composer/behat/behat/src/Behat/Behat/Hook/Tester/HookableScenarioTester.php(74): Behat\Behat\Tester\Runtime\RuntimeScenarioTester->test(Object(Behat\Behat\Context\Environment\InitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), Object(Behat\Gherkin\Node\ScenarioNode), false)
      #17 /drone/src/lib/composer/behat/behat/src/Behat/Behat/EventDispatcher/Tester/EventDispatchingScenarioTester.php(103): Behat\Behat\Hook\Tester\HookableScenarioTester->test(Object(Behat\Behat\Context\Environment\InitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), Object(Behat\Gherkin\Node\ScenarioNode), false)
      #18 /drone/src/lib/composer/behat/behat/src/Behat/Behat/Tester/Runtime/IsolatingScenarioTester.php(69): Behat\Behat\EventDispatcher\Tester\EventDispatchingScenarioTester->test(Object(Behat\Behat\Context\Environment\InitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), Object(Behat\Gherkin\Node\ScenarioNode), false)
      #19 /drone/src/lib/composer/behat/behat/src/Behat/Behat/Tester/Runtime/RuntimeFeatureTester.php(84): Behat\Behat\Tester\Runtime\IsolatingScenarioTester->test(Object(Behat\Behat\Context\Environment\UninitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), Object(Behat\Gherkin\Node\ScenarioNode), false)
      #20 /drone/src/lib/composer/behat/behat/src/Behat/Behat/Hook/Tester/HookableFeatureTester.php(72): Behat\Behat\Tester\Runtime\RuntimeFeatureTester->test(Object(Behat\Behat\Context\Environment\UninitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), false)
      #21 /drone/src/lib/composer/behat/behat/src/Behat/Behat/EventDispatcher/Tester/EventDispatchingFeatureTester.php(71): Behat\Behat\Hook\Tester\HookableFeatureTester->test(Object(Behat\Behat\Context\Environment\UninitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), false)
      #22 /drone/src/lib/composer/behat/behat/src/Behat/Testwork/Tester/Runtime/RuntimeSuiteTester.php(63): Behat\Behat\EventDispatcher\Tester\EventDispatchingFeatureTester->test(Object(Behat\Behat\Context\Environment\UninitializedContextEnvironment), Object(Behat\Gherkin\Node\FeatureNode), false)
      #23 /drone/src/lib/composer/behat/behat/src/Behat/Testwork/Hook/Tester/HookableSuiteTester.php(73): Behat\Testwork\Tester\Runtime\RuntimeSuiteTester->test(Object(Behat\Behat\Context\Environment\UninitializedContextEnvironment), Object(Behat\Testwork\Specification\GroupedSpecificationIterator), false)
      #24 /drone/src/lib/composer/behat/behat/src/Behat/Testwork/EventDispatcher/Tester/EventDispatchingSuiteTester.php(72): Behat\Testwork\Hook\Tester\HookableSuiteTester->test(Object(Behat\Behat\Context\Environment\UninitializedContextEnvironment), Object(Behat\Testwork\Specification\GroupedSpecificationIterator), false)
      #25 /drone/src/lib/composer/behat/behat/src/Behat/Testwork/Tester/Runtime/RuntimeExercise.php(71): Behat\Testwork\EventDispatcher\Tester\EventDispatchingSuiteTester->test(Object(Behat\Behat\Context\Environment\UninitializedContextEnvironment), Object(Behat\Testwork\Specification\GroupedSpecificationIterator), false)
      #26 /drone/src/lib/composer/behat/behat/src/Behat/Testwork/EventDispatcher/Tester/EventDispatchingExercise.php(70): Behat\Testwork\Tester\Runtime\RuntimeExercise->test(Array, false)
      #27 /drone/src/lib/composer/behat/behat/src/Behat/Testwork/Ordering/OrderedExercise.php(80): Behat\Testwork\EventDispatcher\Tester\EventDispatchingExercise->test(Array, false)
      #28 /drone/src/lib/composer/behat/behat/src/Behat/Testwork/Tester/Cli/ExerciseController.php(149): Behat\Testwork\Ordering\OrderedExercise->test(Array, false)
      #29 /drone/src/lib/composer/behat/behat/src/Behat/Testwork/Tester/Cli/ExerciseController.php(108): Behat\Testwork\Tester\Cli\ExerciseController->testSpecifications(Object(Symfony\Component\Console\Input\ArgvInput), Array)
      #30 /drone/src/lib/composer/behat/behat/src/Behat/Testwork/Cli/Command.php(63): Behat\Testwork\Tester\Cli\ExerciseController->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
      #31 /drone/src/lib/composer/symfony/console/Command/Command.php(255): Behat\Testwork\Cli\Command->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
      #32 /drone/src/lib/composer/symfony/console/Application.php(953): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
      #33 /drone/src/lib/composer/symfony/console/Application.php(248): Symfony\Component\Console\Application->doRunCommand(Object(Behat\Testwork\Cli\Command), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
      #34 /drone/src/lib/composer/behat/behat/src/Behat/Testwork/Cli/Application.php(124): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
      #35 /drone/src/lib/composer/symfony/console/Application.php(148): Behat\Testwork\Cli\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
      #36 /drone/src/lib/composer/behat/behat/bin/behat(34): Symfony\Component\Console\Application->run()
      #37 {main}
```

individual-it commented:
@PVince81 looks like users with existing locks cannot be deleted https://github.com/owncloud/core/issues/33769
I haven't noticed locally because I usually use SQLite and that problem does not happen with SQLite

PVince81 commented on apps/files/tests/js/filelockpluginSpec.js:
```diff
@@ -0,0 +1,200 @@
+			locktype: 'write',
+			lockroot: '/owncloud/remote.php/dav/files/currentuser/basepath',
+			depth: 'infinite',
+			timeout: '12345',
```
seems that valid values are "Infinite" or "Second-XYZ" where XYZ is replaced with the number of seconds: https://tools.ietf.org/html/rfc4918

at least we know it's a string

individual-it commented:
@ownclouders rebase

ownclouders commented:
Hey! I'm [GitMate.io](https://gitmate.io)! This pull request is being rebased automatically. Please **DO NOT** push while rebase is in progress or your changes would be lost permanently :warning:

ownclouders commented:
Automated rebase with [GitMate.io](https://gitmate.io) was successful! :tada:

PVince81 commented:
@ownclouders rebase.

expecting https://github.com/owncloud/core/issues/33792 but hopefully other issues related to locking will be solved

ownclouders commented:
Hey! I'm [GitMate.io](https://gitmate.io)! This pull request is being rebased automatically. Please **DO NOT** push while rebase is in progress or your changes would be lost permanently :warning:

ownclouders commented:
Automated rebase with [GitMate.io](https://gitmate.io) was successful! :tada:

PVince81 commented:
- apiMain failed as in core.
- I've restarted apiSharees test due to Github issue.
- the locking tests all green :tada:

PVince81 commented:
ok I just expanded the Github PR as it was hiding all the bug reports... I'll have a look tomorrow

PVince81 commented:
- [ ] bug: creating a lock of a non-existing file will create such file with 0 bytes in oc_filecache

@DeepDiver1975 is this intended ?

phil-davis commented:
@ownclouders rebase

failing crud in master has been fixed, so get a clean CI here

ownclouders commented:
Hey! I'm [GitMate.io](https://gitmate.io)! This pull request is being rebased automatically. Please **DO NOT** push while rebase is in progress or your changes would be lost permanently :warning:

ownclouders commented:
Automated rebase with [GitMate.io](https://gitmate.io) was successful! :tada:

PVince81 commented:
- [ ] bug: admin cannot unlock ??

PVince81 commented:
raised https://github.com/owncloud/core/issues/33822 for the issue when locking as the email user

individual-it commented:
I've adjusted the test expectations to the current error messages

@PVince81 should I get rid of the test code related to unlocking, or just skip the tests?

phil-davis commented on tests/acceptance/features/webUIWebdavLocks/locks.feature:
```diff
@@ -417,7 +417,7 @@ Feature: Locks
     When the user deletes file "lorem.txt" using the webUI
     Then notifications should be displayed on the webUI with the text
-        | file "lorem.txt" is locked and cannot be deleted. |
+        | The The file "lorem.txt" is locked and cannot be deleted. |
```
@individual-it repeated ``The The`` is surprising?

phil-davis commented on tests/acceptance/features/webUIWebdavLocks/locks.feature:
```diff
@@ -545,7 +545,7 @@ Feature: Locks
     And the user deletes folder "lorem.txt" using the webUI
     Then notifications should be displayed on the webUI with the text
-      | file "lorem.txt" is locked and cannot be deleted. |
+      | The The file "lorem.txt" is locked and cannot be deleted. |
```
@individual-it repeated ``The The`` is surprising?

PVince81 commented:
> I've adjusted the test expectations to the current error messages
> 
> @PVince81 should I get rid of the test code related to unlocking, or just skip the tests?

let's wait a little more. I'll have a chat with @DeepDiver1975 on Monday to reclarify the RFC

PVince81 commented:
As discussed with @DeepDiver1975 I've brought back the unlock button. We always display it as the frontend doesn't have enough info to decide whether unlocking is possible.

Whenever the user clicks it, the backend will decide whether the user is able to unlock it. In case of 403, a message will tell the user to contact the lock owner and display the owner string in a notification.

Otherwise the lock is correctly removed.

PVince81 commented:
raised https://github.com/owncloud/core/issues/33847 for the backend issue about unsharing from self

PVince81 commented:
> bug: lock information are not displayed on the "Shared with You" or "Shared with others" page

we don't have this information as the list is built not by Webdav but through the OCS Share API.
it would require extending the OCS Share API to also include the lock info... something for later.

PVince81 commented:
raised https://github.com/owncloud/core/issues/33848 which I found while investigating the overwrite issue

PVince81 commented:
fixed file upload overwrite lock error to be better than "Locked".

I've addressed all issues. There's only one left to raise separately after merge related to "Shared with you".

@individual-it do you have any outstanding tasks ? (would be good to list as checkboxes)

Once done we can manually squash commits into logical blocks.

individual-it commented:
@ownclouders rebase

ownclouders commented:
Hey! I'm [GitMate.io](https://gitmate.io)! This pull request is being rebased automatically. Please **DO NOT** push while rebase is in progress or your changes would be lost permanently :warning:

ownclouders commented:
Automated rebase with [GitMate.io](https://gitmate.io) was successful! :tada:

individual-it commented:
@PVince81 only two tests are left failing `creating a subfolder structure that is the same as the structure of a declined & locked share`
the linked issue in https://github.com/owncloud/core/pull/32250#issuecomment-434615887 has IMHO nothing to do with the problem

PVince81 commented:
@individual-it I've moved "bug: creating a folder structure with the same names as a shared and declined that is locked also locks the local structure" to a separate backend bug: #33885

please disable the test accordingly for now

PVince81 commented:
- [ ] double check whether oc_persistentlocks supports depth infinity as I see "int" used...

PVince81 commented:
I've squashed all this into three commits, one for each of us.

@individual-it I've disabled the test for https://github.com/owncloud/core/issues/33885

Expecting all green now.

DeepDiver1975 commented:
Fine by me :+1:
